### PR TITLE
100-continue functional tests

### LINKED ISSUE #617: Test 100-continue scenarios
https://github.com/microsoft/reverse-proxy/blob/37ed6a93f3ea54cc412471ca85ac9b3e38aca749/src/ReverseProxy/Service/Proxy/HttpProxy.cs#L182-L186

That code makes an assumption that request bodies are always read, which is likely false in the case of 100-continues. That check is also one of the few places where it would be problematic if a transform replaced the HttpContent object, the Started property wouldn't be available.

This exception should probably be removed.

### PR DESCRIPTION
**Pull request by alnikola**
It adds functional tests for the scenarios listed below. They are based on [this list](https://github.com/microsoft/reverse-proxy/issues/617#issuecomment-852243785), but modified to be compatible with the actual HttpClient behavior.

- A. HTTP/1.1 POST with a Content-Length is proxied over HTTP/1.1, destination responds with a 100 Continue, the proxy responds with a 100 Continue, the data is uploaded, there's a 200 response.
- B. HTTP/1.1 POST with a Transfer-Encoding is proxied over HTTP/1.1, destination responds with a 100 Continue, the proxy responds with a 100 Continue, the data is uploaded, there's a 200 response.
- C. A & B proxied over HTTP/2
- D. HTTP/1.1 POST with a Content-Length is proxied over HTTP/1.1, destination responds with a 4xx, the data is not uploaded.
- E. D but with Transfer-Encoding.
- F. HTTP/1.1 POST with a Content-Length is proxied over HTTP/2, destination responds with a 4xx and a RESET, the data is not uploaded.
- G. HTTP/2 POST with a Content-Length is proxied over HTTP/2, destination responds with a 4xx and a RESET, the proxy app sends the 4xx and a RESET to cancel the body.

Fixes #617

### CODE REVIEW

File: test/ReverseProxy.FunctionalTests/Expect100ContinueTests.cs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+                        headerTcs.SetException(new Exception("Missing 'Expect' header in request"));
+                    }
+
+                    if (destResponseCode == 200)
+                    {
+                        // 100 response code is sent automatically on reading Body.
+                        await ReadContent(context, bodyTcs, contentString);
+                    }
+                    context.Response.StatusCode = destResponseCode;
+                    await context.Response.CompleteAsync();
```
Comment (Tratcher) [inline test/ReverseProxy.FunctionalTests/Expect100ContinueTests.cs:58]: Not needed, the response is automatically sent when this method exits.
```suggestion
```

File: test/ReverseProxy.FunctionalTests/Expect100ContinueTests.cs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            var content = Encoding.UTF8.GetBytes(contentString);
+            using var contentStream = new MemoryStream(content);
+            message.Content = new StreamContent(contentStream);
+            if (useContentLength)
+            {
+                message.Content.Headers.ContentLength = content.Length;
+            }
+            else
+            {
+                message.Content.Headers.ContentEncoding.Add("chunked");
```
Comment (Tratcher) [inline test/ReverseProxy.FunctionalTests/Expect100ContinueTests.cs:145]:
```suggestion
                message.Headers.TransferEncodingChunked = true;
```

File: test/ReverseProxy.FunctionalTests/Expect100ContinueTests.cs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            }
+            catch (Exception e)
+            {
+                bodyTcs.SetException(e);
+            }
+        }
+
+        private async Task ProcessHttpRequest(Uri proxyHostUri, HttpProtocols protocol, string contentString, bool useContentLength, int expectedCode)
+        {
+            using var handler = new SocketsHttpHandler() { Expect100ContinueTimeout = TimeSpan.FromSeconds(60) };
```
[the remark this answers is not in the document]
Comment (Tratcher) [reply test/ReverseProxy.FunctionalTests/Expect100ContinueTests.cs:125]: You'll want to set a long 100-continue timeout on the proxy's handler instance as well.

**Review by Tratcher — COMMENTED** [review]
I'm a little surprised the tests are all passing, I was expecting some errors for the 4xx responses.

One variant I did not consider before: Do these scenarios also work if there is a response body? I'm mainly concerned about the 4xx scenario where resets are being sent, I don't know if they'll interfere with the response body.

File: test/ReverseProxy.FunctionalTests/Expect100ContinueTests.cs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+                configTransformer: (c, r) =>
+                {
+                    c = c with
+                    {
+                        HttpRequest = new Forwarder.ForwarderRequestConfig
+                        {
+                            Version = destProtocol == HttpProtocols.Http2 ? HttpVersion.Version20 : HttpVersion.Version11,
+#if NET
+                            VersionPolicy = HttpVersionPolicy.RequestVersionExact
+#endif
```
[the remark this answers is not in the document]
Comment (alnikola) [reply test/ReverseProxy.FunctionalTests/Expect100ContinueTests.cs]: It was a mistake. It's better to use HTTPS to negotiate protocol version. Fixed.

File: test/ReverseProxy.FunctionalTests/Expect100ContinueTests.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (alnikola) [reply test/ReverseProxy.FunctionalTests/Expect100ContinueTests.cs:126]: Done.

**Comment by alnikola** [comment]
Obsoleted by #1185

**Pull request closed without merging**